### ISSUE TITLE
HttpClient ignores Timeout and cancellation token for slow response body reads on Windows

### ISSUE THREAD

**Issue by davidsh** [dotnet/corefx#8663]
(Split from #8627 which contained multiple related bugs)

This bug is in the core logic for HttpClient, HttpContent classes and will affect both Windows and *Nix.

The following code demonstrates the problem with all of the HttpClient *Async() methods when set to buffer in the response body.

This test server will return response headers quickly. But will take 15 seconds to return response body. In these examples, we call GetAsync() and include the option (default) to read in (buffer into memory) the response body as well as response headers.
Each test uses a different method to cancel the operation after 1 second. Either the .Timeout property or a cancellation token. 

In both case, we request cancellation after 1 second. But the GetAsync() task doesn't cancel until the 15 seconds has elapsed.

This pattern works on .NET Framework (Desktop). Looking at the Desktop code, it is surprising that it works because the cancellation token passed into `HttpClient.GetAsync` is not actually propagated down thru the stack. The cancellation logic only works on Desktop because of a special internal method used by Desktop HTTP stack (HttpWebRequest). Since we have a different HTTP stack for CoreFx, this doesn't work.

```c#
[ActiveIssue(8663)]
[Fact]
public async Task GetIncludesReadingResponseBody_CancelUsingTimeout_TaskCanceledInShortTime()
{
    var stopwatch = new Stopwatch();
    stopwatch.Start();

    using (var client = new HttpClient())
    {
        client.Timeout = new TimeSpan(0, 0, 1);

        Task<HttpResponseMessage> getResponse =
            client.GetAsync(s_fastHeadersSlowBodyServer, HttpCompletionOption.ResponseContentRead);

        stopwatch.Restart();
        await Assert.ThrowsAsync<TaskCanceledException>(
            () => getResponse);
        stopwatch.Stop();
        _output.WriteLine("GetAsync() completed at: {0}", stopwatch.Elapsed.ToString());
        Assert.True(stopwatch.Elapsed < new TimeSpan(0,0,3), "Elapsed time should be short");
    }
}

[ActiveIssue(8663)]
[Fact]
public async Task GetIncludesReadingResponseBody_CancelUsingToken_TaskCanceledInShortTime()
{
    var stopwatch = new Stopwatch();
    stopwatch.Start();

    var cts = new CancellationTokenSource();
    using (var client = new HttpClient())
    {
        Task<HttpResponseMessage> getResponse =
            client.GetAsync(s_fastHeadersSlowBodyServer, HttpCompletionOption.ResponseContentRead, cts.Token);

        Task ignore = Task.Delay(new TimeSpan(0, 0, 1)).ContinueWith(_ =>
        {
            _output.WriteLine("Calling cts.Cancel() at: {0}", stopwatch.Elapsed.ToString());
            cts.Cancel();
        });

        stopwatch.Restart();
        await Assert.ThrowsAsync<TaskCanceledException>(
            () => getResponse);
        stopwatch.Stop();
        _output.WriteLine("GetAsync() completed at: {0}", stopwatch.Elapsed.ToString());
        Assert.True(stopwatch.Elapsed < new TimeSpan(0,0,3), "Elapsed time should be short");
    }
}
```

**Comment by davidsh** [comment]
Even when #8662 is fixed, there still needs to be work done to propagate the cancellation token to the response stream objects.

**Comment by davidsh** [comment]
Moving post RTM.

**Comment by joshfree** [comment]
Fixed by https://github.com/dotnet/corefx/pull/9110 for /release/1.0.0

**Comment by davidsh** [comment]
Fixed with PR #9446.
